libpromises/evalfunction.c

```diff
@@ -8668,7 +8685,11 @@ static FnCallResult FnCallIsReadable(ARG_UNUSED EvalContext *const ctx,
                 "Read operation timed out, exceeded %ld seconds.", path,
                 timeout);
 
+#ifdef HAVE_PTHREAD_CANCEL
             ret = pthread_cancel(thread_data.thread);
+#else
+            ret = pthread_kill(thread_data.thread, SIGUSR2);
```

Comment:
@larsewi you had mentioned in my previous PR [comment](https://github.com/cfengine/core/pull/5574#discussion_r1670902889)

> Is it safe to join a "killed" thread? Will this call the signal handler installed by the agent. If so, is this fine? Also the man pthread_kill(3) says:

And so here I am using the downstream patch which sends a kill signal and adds a signal handler to which then calls `pthread_exit` which is the "other option" of how the joined thread could be terminated at the appropriate timeout moment.

> After a canceled thread has terminated, a join with that thread
       using [pthread_join(3)](https://www.man7.org/linux/man-pages/man3/pthread_join.3.html) obtains PTHREAD_CANCELED as the thread's
       exit status.  (Joining with a thread is the only way to know that
       cancelation has completed.)

from https://www.man7.org/linux/man-pages/man3/pthread_cancel.3.html